domain/network/state/space.go

```diff
@@ -42,85 +41,92 @@ func (st *State) AddSpace(
 		return errors.Trace(err)
 	}
 
-	insertSpaceStmt := `
-INSERT INTO space (uuid, name)
-VALUES (?, ?)`
+	insertSpaceStmt, err := sqlair.Prepare(
+		`INSERT INTO space (uuid, name) VALUES ($Space.uuid, $Space.name)`, Space{},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	insertProviderStmt := `
-INSERT INTO provider_space (provider_id, space_uuid)
-VALUES (?, ?)`
+	insertProviderStmt, err := sqlair.Prepare(
+		`INSERT INTO provider_space (provider_id, space_uuid) VALUES ($ProviderSpace.provider_id, $ProviderSpace.space_uuid)`,
+		ProviderSpace{},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	subnetBinds, subnetVals := database.SliceToPlaceholder(subnetIDs)
-	findFanSubnetsStmt := fmt.Sprintf(`
-SELECT subject_subnet_uuid
+	findFanSubnetsStmt, err := sqlair.Prepare(`
+SELECT subject_subnet_uuid AS &Subnet.uuid
 FROM   subnet_association
-WHERE  association_type_id = 0 AND associated_subnet_uuid IN (%s)`, subnetBinds)
+WHERE  association_type_id = 0 AND associated_subnet_uuid IN ($S[:])`, sqlair.S{}, Subnet{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	checkInputSubnetsStmt := fmt.Sprintf(`
-SELECT uuid
+	checkInputSubnetsStmt, err := sqlair.Prepare(`
+SELECT &Subnet.uuid
 FROM   subnet
 JOIN   subnet_type
 ON     subnet.subnet_type_id = subnet_type.id
-WHERE  subnet_type.is_space_settable = FALSE AND subnet.uuid IN (%s)`, subnetBinds)
+WHERE  subnet_type.is_space_settable = FALSE AND subnet.uuid IN ($S[:])`, sqlair.S{}, Subnet{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		// We must first check on the provided subnet ids to validate
 		// that are of a type on which the space can be set.
-		nonSettableSubnets, err := tx.QueryContext(ctx, checkInputSubnetsStmt, subnetVals...)
+
+		subnetIDsInS := sqlair.S{}
+		for _, sid := range subnetIDs {
+			subnetIDsInS = append(subnetIDsInS, sid)
+		}
```

Comment:
Move this outside of the `db.Txn` function call. The `Txn` can be retried multiple times, so we only want to do things once if we can help it.

Reply:
Moved to outside Txn.